google-aec: Component unit test

lkoenig opened the pull request:
This unit tests ensure all function got called. It does not verify audio
processing behaviour but allow testing for memory allocation/free and
hence tests leakage.

Signed-off-by: Lionel Koenig <lionelk@google.com>

lgirdwood reviewed (commented):
@singalsu any comment ?

cujomalainey commented on test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:
```diff
@@ -0,0 +1,191 @@
+#include <sof/audio/component_ext.h>
+#include <sof/lib/uuid.h>
+
+#define FRAME_SIZE 48
```
i thought we do 10ms?

cujomalainey commented on test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:
```diff
@@ -0,0 +1,191 @@
+
+	comp_trigger(google_rtc_audio_processing_state->dev,
+		     COMP_TRIGGER_PREPARE);
+	comp_prepare(google_rtc_audio_processing_state->dev);
```
do we need to check these return codes?

lkoenig commented:
We do 10ms. It should work with almost any size. The processing will only occure for every 10ms of audio.
The component should not have to care how much the period size is.

lkoenig commented:
I added checks for the return codes.
I also added some supported use-case in term of number of channels and period size.

This test won't work on main due to the UUID issue I mentioned earlier: in `src/include/sof/lib/uuid.h` the macros are defined to return `NULL`when `CONFIG_LIB` is defined. Which will cause an issue in the test setup when trying to construct the component. @singalsu I was wondering if you had an idea here or if I am doing something wrong ?

sys-pt1s commented:
Can one of the admins verify this patch?

cujomalainey commented:
@lkoenig that sounds like something I am trying to undo. I need UUIDs in CONFIG_LIB for the fuzzer.

singalsu commented:
@lkoenig I was OOO for two weeks, have you been able to solve this? I haven't done cmocka test cases for components after UUID introduction so I may have well missed this.

lkoenig commented:
@cujomalainey let me know when you found a solution so I cna update this PR.
@singalsu I haven't solve it yet (counting on @cujomalainey) to figure something smart up.

lgirdwood reviewed (approved):
I think the UT looks fine, @singalsu any comments ?

lgirdwood commented:
> It depends on how we mock the build if we add back the linker script or not to do dynamic registration. @lgirdwood do you know the plan here?

@singalsu has some work to look at the changes needed to run actual processing tests as part of CI using testbench, this would include the linkage. We have made some Testbench cleanup and improvement, but not sure if everything is in place yet to run AEC testing ?

singalsu commented on test/cmocka/src/audio/google_rtc_audio_processing/google_rtc_audio_processing_process.c:
```diff
@@ -0,0 +1,255 @@
+
+static void verify_s16_to_s16(struct comp_dev *dev, struct comp_buffer *mic,
+			      struct comp_buffer *ref, struct comp_buffer *out)
+{
```
The magic is missing :)

cujomalainey replied:
@lkoenig just want to confirm this is still not running any actual data validation

singalsu reviewed (commented):
Looks very promising, thanks!

singalsu commented:
> @singalsu has some work to look at the changes needed to run actual processing tests as part of CI using testbench, this would include the linkage. We have made some Testbench cleanup and improvement, but not sure if everything is in place yet to run AEC testing ?

I have been able to build and run the mockup version of this AEC in testbench. The test topologies for such purpose needs some thinking but I could make a quick draft PR to show how and how to get CI to test it. I have also work done to run all current process components in testbench in CI but it needs polishing and a new PR. It was earlier part of other PR and not really in that PR context so it was postponed.

lkoenig commented:
Sorry for the long delay, I was busy on other projects.

cujomalainey commented:
Looks LGTM to me, just curious if we will do any data validation, if not then I will mark as approved

lkoenig commented:
One think we can do is check that we have audio energy on the output when energy is fed on the input. But event that is not really reliable as the AEC might not be that transparent. That would work for the mock though. If you feel this is something worth pursuing I can add a very small check for that.

lkoenig commented:
One think we can do is check that we have audio energy on the output when energy is fed on the input. But event that is not really reliable as the AEC might not be that transparent. That would work for the mock though. If you feel this is something worth pursuing I can add a very small check for that.

cujomalainey commented:
Lets sync on this, I just want to make sure this gets the best value for its effort.

kv2019i commented:
Converting to draft, no activity in 3 months.

cujomalainey commented:
I think this can be closed out until we get further updates
